docs/development.md

```diff
@@ -18,6 +18,80 @@ sudo apt-get install libblkid-dev
+
+### Validation and failures
+
+Ignition is a low-level tool and does not attempt to prevent configs from doing unreasonable things.
+
+- Config validation should fail when it's clear from inspection that Ignition will be unable to perform the requested provisioning.  This must occur when the config is contradictory (e.g. specifies the same filesystem object as both a file and a directory) and thus would break the declarativeness of the spec, and may occur where the problem will produce a routine error at runtime (e.g. invalid arguments to `mkfs`).
+- Ignition must fail at runtime when it is dynamically unable to perform the requested provisioning.
+- If a config is implementable but will render the system non-functional, Ignition should execute the config anyway.  Any user-friendly detection of unreasonable configs should happen in Butane.
+
+### Platforms
+
+Platform providers should continue retrying config fetch until it's clear whether the user has provided an Ignition config.  If Ignition eventually timed out when fetching a config, then a slow network device or block device (on a very large machine or a heavily-loaded VM host) could cause Ignition to prematurely fail, or to continue booting without applying the specified config.
+
+Platform providers must allow the user not to provide a config, e.g. to boot an exploratory OS instance and use Afterburn to inject SSH keys.
+
+Ignition must never read from config providers that aren't under the control of the platform, since this could allow config injection from unintended sources.  For example, `169.254.169.254` is a link-local address and could easily be spoofed on platforms that don't specially handle that address.  As a corollary, the platform ID must always be explicitly set by the OS image, never guessed.
+
+### Network access
+
+All config fields that cause network accesses, directly or indirectly, should be added to the `fetch-offline` needs-net detector.
+
+Any network accesses performed by subprocesses should be mimicked by Ignition before starting the subprocess.  This ensures that Ignition's retry logic is used, so Ignition doesn't improperly fail if the network is still coming up.
+
+### Execution stages
+
+Ignition execution is divided into stages to allow other OS functionality to run in the middle of provisioning:
+
+1. `fetch-offline` stage
+1. OS enables networking if required
+1. `fetch` stage
+1. OS examines fetched config and e.g. copies root filesystem contents to RAM
+1. `disks` stage
+1. OS mounts root filesystem
+1. `mount` stage
+1. OS does any preprocessing of configured filesystems, including copying root filesystem contents back
+1. `files` stage
+1. OS does any postprocessing of provisioned system
+1. `umount` stage
```

Comment:
Would be nice if the description of each stage is a sub-bullet. Right now it looks a bit awkward:

![Screenshot 2023-08-11 at 10-33-48 Development](https://github.com/coreos/ignition/assets/4530030/c8db4ec1-9342-4733-b1b4-1c5c732d7bdf)



Also would be nice if we could add a description for the umount stage like all the others even though it's a bit superfluous.

Reply:
Those aren't descriptions of the stages; they're descriptions of what happens between the stages.  So it's rendering correctly.

Reply:
ahh - I see. I should have read more closely.